File: src/lib/subscriptions.ts
import type { Cadence, Subscription } from '@/db/types';

/**
 * What a subscription costs and when it renews next.
 *
 * All of it is arithmetic on two fields — a cadence and an anchor date — and
 * all of it is here rather than in a screen, because every one of these
 * calculations has an off-by-one in it somewhere and a wrong renewal date is
 * indistinguishable from a right one until the money leaves.
 *
 * WHY AN ANCHOR AND NOT A DAY NUMBER
 *
 * "Which day of the month does it renew" is the obvious model and it can only
 * describe monthly plans. Amazon Prime and YouTube Premium are commonly annual;
 * a lot of services push an annual price precisely because people forget it. A
 * yearly plan stored as a day-of-month becomes a monthly charge, and the
 * dashboard total is then wrong by a factor of twelve — quietly, and in the
 * flattering direction.
 *
 * So: an anchor date, which is one real renewal (usually the first), and a
 * cadence. The day of the month falls out of the anchor for the monthly case,
 * and the annual case is representable.
 */

const DAY = 86_400_000;

/** How many months one period spans. Weekly is handled separately. */
const MONTHS_PER: Record<Exclude<Cadence, 'weekly'>, number> = {
  monthly: 1,
  quarterly: 3,
  yearly: 12,
};

/** Periods in a year, for normalising to a monthly figure. */
const PER_YEAR: Record<Cadence, number> = {
  weekly: 365.25 / 7,
  monthly: 12,
  quarterly: 4,
  yearly: 1,
};

export const CADENCES: Cadence[] = ['weekly', 'monthly', 'quarterly', 'yearly'];

export const CADENCE_LABEL: Record<Cadence, string> = {
  weekly: 'Weekly',
  monthly: 'Monthly',
  quarterly: 'Quarterly',
  yearly: 'Yearly',
};

/** "a month" / "a year" — for "$12.99 a month". */
export const CADENCE_PER: Record<Cadence, string> = {
  weekly: 'a week',
  monthly: 'a month',
  quarterly: 'a quarter',
  yearly: 'a year',
};

/** Local midnight, so comparisons are about days rather than hours. */
export function startOfDay(d: Date): Date {
  return new Date(d.getFullYear(), d.getMonth(), d.getDate());
}

export function parseAnchor(iso: string): Date | null {
  const m = /^(\d{4})-(\d{2})-(\d{2})$/.exec(iso.trim());
  if (!m) return null;
  const [y, mo, da] = [Number(m[1]), Number(m[2]), Number(m[3])];
  const d = new Date(y, mo - 1, da);
  // Rejects 2026-02-31 and friends, which `new Date` silently rolls forward.
  return d.getFullYear() === y && d.getMonth() === mo - 1 && d.getDate() === da ? d : null;
}

/**
 * Add months, clamping to the end of the target month.
 *
 * A subscription anchored on the 31st renews on the 30th in April and the 28th
 * in February — that is what the card issuer does, and rolling forward into
 * the 1st of the next month instead would put the renewal in the wrong month
 * on the calendar and quietly shift every subsequent date by one.
 */
export function addMonthsClamped(from: Date, months: number): Date {
  const y = from.getFullYear();
  const m = from.getMonth() + months;
  const lastDay = new Date(y, m + 1, 0).getDate();
  return new Date(y, m, Math.min(from.getDate(), lastDay));
}

/**
 * The next time this renews, on or after `now`.
 *
 * Stepping period by period rather than computing a count, because clamped
 * months don't divide: an anchor on the 31st visits the 28th of February and
 * must come back to the 31st in March, which only works if each step is taken
 * from the original anchor rather than from the previous result.
 */
export function nextRenewal(sub: Pick<Subscription, 'cadence' | 'anchorDate'>, now = new Date()): Date | null {
  const anchor = parseAnchor(sub.anchorDate);
  if (!anchor) return null;

  const today = startOfDay(now);
  if (anchor >= today) return anchor;

  if (sub.cadence === 'weekly') {
    const weeks = Math.ceil((today.getTime() - anchor.getTime()) / (7 * DAY));
    return new Date(anchor.getTime() + weeks * 7 * DAY);
  }

  const step = MONTHS_PER[sub.cadence];
  // Start from a floor estimate and walk, so a clamped month can't strand us.
  const months =
    (today.getFullYear() - anchor.getFullYear()) * 12 + (today.getMonth() - anchor.getMonth());
  let periods = Math.max(0, Math.floor(months / step));
  for (let guard = 0; guard < 500; guard++) {
    const at = addMonthsClamped(anchor, periods * step);
    if (at >= today) return at;
    periods++;
  }
  return null;
}

/** Whole days until the next renewal. 0 means today. */
export function daysUntilRenewal(
  sub: Pick<Subscription, 'cadence' | 'anchorDate'>,
  now = new Date(),
): number | null {
  const at = nextRenewal(sub, now);
  if (!at) return null;
  return Math.round((at.getTime() - startOfDay(now).getTime()) / DAY);
}

/**
 * What this costs per month, in cents.
 *
 * Yearly divided by twelve, weekly multiplied by 52.18 — the point of the
 * figure is comparison and a monthly total, so a weekly plan has to be
 * expressed as its true monthly average rather than four weeks, which
 * undercounts by about 8%.
 */
export function monthlyCents(sub: Pick<Subscription, 'cadence' | 'amountCents'>): number {
  if (!Number.isFinite(sub.amountCents) || sub.amountCents <= 0) return 0;
  return Math.round((sub.amountCents * PER_YEAR[sub.cadence]) / 12);
}

export function totalMonthlyCents(subs: Pick<Subscription, 'cadence' | 'amountCents'>[]): number {
  return subs.reduce((sum, s) => sum + monthlyCents(s), 0);
}

/** Yearly spend, from the same normalisation. */
export function totalYearlyCents(subs: Pick<Subscription, 'cadence' | 'amountCents'>[]): number {
  return subs.reduce((sum, s) => sum + Math.round(s.amountCents * PER_YEAR[s.cadence]), 0);
}

/* --------------------------------------------------------------- reminders */

/** Days before renewal a reminder can be set for. `0` means none. */
export const REMIND_CHOICES = [0, 1, 3, 7] as const;
export type RemindDays = (typeof REMIND_CHOICES)[number];

/**
 * Whether this subscription wants to be mentioned right now.
 *
 * "Right now" means the next time the app is opened, because that is the only
 * moment it can say anything: there is no server and nothing runs while the
 * app is closed. Everything about the wording around this setting has to be
 * honest about that — a reminder that only appears once you've opened the app
 * is a note, and calling it an alert would be a promise the app can't keep.
 */
export function reminderDue(sub: Subscription, now = new Date()): boolean {
  if (!sub.remindDays) return false;
  const days = daysUntilRenewal(sub, now);
  if (days === null) return false;
  return days >= 0 && days <= sub.remindDays;
}

/** Everything that wants mentioning, soonest first. */
export function dueReminders(subs: Subscription[], now = new Date()): Subscription[] {
  return subs
    .filter((s) => reminderDue(s, now))
    .sort((a, b) => (daysUntilRenewal(a, now) ?? 0) - (daysUntilRenewal(b, now) ?? 0));
}

/* ----------------------------------------------------------- the calendar */

export interface RenewalDay {
  /** 1..31 */
  day: number;
  subs: Subscription[];
}

/**
 * Which day of a given month each subscription lands on.
 *
 * Computed by walking the month rather than reading the anchor's day number,
 * so a weekly plan appears four or five times and a yearly one appears in only
 * one month of the twelve — which is the whole reason to draw a calendar
 * rather than a list of day numbers.
 */
export function renewalsInMonth(
  subs: Subscription[],
  year: number,
  month: number,
  now = new Date(),
): RenewalDay[] {
  const days = new Date(year, month + 1, 0).getDate();
  const out: RenewalDay[] = [];

  for (let day = 1; day <= days; day++) {
    const date = new Date(year, month, day);
    const on = subs.filter((s) => {
      const at = nextRenewal(s, date <= startOfDay(now) ? now : date);
      return at ? at.getFullYear() === year && at.getMonth() === month && at.getDate() === day : false;
    });
    if (on.length) out.push({ day, subs: on });
  }
  return out;
}

/** "Renews today" / "Renews tomorrow" / "Renews in 6 days". */
export function renewalLabel(days: number | null): string {
  if (days === null) return 'No renewal date';
  if (days <= 0) return 'Renews today';
  if (days === 1) return 'Renews tomorrow';
  return `Renews in ${days} days`;
}

/** "3rd", for the compact row under a logo. */
export function ordinal(n: number): string {
  const rest = n % 100;
  if (rest >= 11 && rest <= 13) return `${n}th`;
  return `${n}${['th', 'st', 'nd', 'rd'][n % 10] ?? 'th'}`;
}

File: mobile/lib/logic/subscriptions.dart
/// What a subscription costs and when it renews next.
///
/// Translated from `src/lib/subscriptions.ts`.
///
/// All of it is arithmetic on two fields — a cadence and an anchor date — and
/// all of it is here rather than in a screen, because every one of these
/// calculations has an off-by-one in it somewhere and a wrong renewal date is
/// indistinguishable from a right one until the money leaves.
///
/// ── Why an anchor and not a day number ────────────────────────────────────
/// "Which day of the month does it renew" is the obvious model and it can only
/// describe monthly plans. A yearly plan stored as a day-of-month becomes a
/// monthly charge, and the dashboard total is then wrong by a factor of twelve
/// — quietly, and in the flattering direction.
library;

import '../models/subscription.dart';
import 'dates.dart';

/// How many months one period spans. Weekly is handled separately.
const Map<Cadence, int> _monthsPer = {
  Cadence.monthly: 1,
  Cadence.quarterly: 3,
  Cadence.yearly: 12,
};

/// Periods in a year, for normalising to a monthly figure.
const Map<Cadence, double> _perYear = {
  Cadence.weekly: 365.25 / 7,
  Cadence.monthly: 12,
  Cadence.quarterly: 4,
  Cadence.yearly: 1,
};

const Map<Cadence, String> cadenceLabel = {
  Cadence.weekly: 'Weekly',
  Cadence.monthly: 'Monthly',
  Cadence.quarterly: 'Quarterly',
  Cadence.yearly: 'Yearly',
};

/// "a month" / "a year" — for "$12.99 a month".
const Map<Cadence, String> cadencePer = {
  Cadence.weekly: 'a week',
  Cadence.monthly: 'a month',
  Cadence.quarterly: 'a quarter',
  Cadence.yearly: 'a year',
};

/// The next time this renews, on or after `now`.
///
/// Stepping period by period rather than computing a count, because clamped
/// months do not divide: an anchor on the 31st visits the 28th of February and
/// must come back to the 31st in March, which only works if each step is taken
/// from the ORIGINAL anchor rather than from the previous result.
DateTime? nextRenewal(Subscription sub, [DateTime? now]) {
  final anchor = parseDate(sub.anchorDate);
  if (anchor == null) return null;

  final today = startOfDay(now ?? DateTime.now());
  if (!anchor.isBefore(today)) return anchor;

  if (sub.cadence == Cadence.weekly) {
    // Counted in days and stepped by the calendar. Done in milliseconds this
    // silently moved every weekly renewal one day earlier for the winter half
    // of the year — see the note at the top of dates.dart.
    final weeks = (daysBetween(anchor, today) / 7).ceil();
    return addDays(anchor, weeks * 7);
  }

  final step = _monthsPer[sub.cadence]!;

  // Start from a floor estimate and walk, so a clamped month cannot strand us.
  final months = (today.year - anchor.year) * 12 + (today.month - anchor.month);
  var periods = months ~/ step;
  if (periods < 0) periods = 0;

  for (var guard = 0; guard < 500; guard++) {
    final at = addMonthsClamped(anchor, periods * step);
    if (!at.isBefore(today)) return at;
    periods++;
  }
  return null;
}

/// Whole days until the next renewal. 0 means today.
int? daysUntilRenewal(Subscription sub, [DateTime? now]) {
  final at = nextRenewal(sub, now);
  if (at == null) return null;
  return daysBetween(startOfDay(now ?? DateTime.now()), at);
}

/// What this costs per month, in cents.
///
/// Yearly divided by twelve, weekly multiplied by 52.18 — the point of the
/// figure is comparison and a monthly total, so a weekly plan has to be
/// expressed as its true monthly average rather than four weeks, which
/// undercounts by about 8%.
int monthlyCents(Subscription sub) {
  if (sub.amountCents <= 0) return 0;
  return (sub.amountCents * _perYear[sub.cadence]! / 12).round();
}

int totalMonthlyCents(List<Subscription> subs) =>
    subs.fold(0, (sum, s) => sum + monthlyCents(s));

/// Yearly spend, from the same normalisation.
int totalYearlyCents(List<Subscription> subs) => subs.fold(
      0,
      (sum, s) => sum + (s.amountCents * _perYear[s.cadence]!).round(),
    );

/// The single largest recurring charge, normalised to a month.
Subscription? biggest(List<Subscription> subs) {
  if (subs.isEmpty) return null;
  final sorted = [...subs]..sort((a, b) => monthlyCents(b) - monthlyCents(a));
  return sorted.first;
}

class DueWithin {
  const DueWithin(this.count, this.cents);
  final int count;
  final int cents;
}

/// What actually leaves your account in the next `days`.
///
/// Not a normalised figure — the real charges, on their real dates. The
/// monthly total answers "what is this costing me", which is a question about
/// the year; this answers "what is about to happen", which is a question about
/// Thursday.
DueWithin dueWithin(List<Subscription> subs, int days, [DateTime? now]) {
  var count = 0;
  var cents = 0;
  for (final s in subs) {
    final left = daysUntilRenewal(s, now);
    if (left == null || left < 0 || left > days) continue;
    count++;
    cents += s.amountCents;
  }
  return DueWithin(count, cents);
}

/// The monthly total, per day.
///
/// The same money said in the unit people actually feel. "$94 a month" is a
/// line on a statement; "about $3 a day" is a coffee, and it is the framing
/// that makes somebody look at the list.
double dailyCents(List<Subscription> subs) =>
    ((totalYearlyCents(subs) / 365.25) * 100).round() / 100;

/* --------------------------------------------------------------- reminders */

/// Days before renewal a reminder can be set for. `0` means none.
const List<int> remindChoices = [0, 1, 3, 7];

/// Whether this subscription wants to be mentioned right now.
///
/// "Right now" means the next time the app is opened, because on the web that
/// was the only moment it could say anything: there was no server and nothing
/// ran while the app was closed. Every word around this setting was written to
/// be honest about that.
///
/// **This is one of the promises the port can finally keep.** A native app has
/// a real scheduler, so in phase 5 this becomes a notification that arrives
/// whether the app is open or not. The predicate does not change — what
/// changes is that something other than a screen render can ask it.
///
/// `remindDays` null or 0 both mean no reminder, which is the default: nine
/// monthly services would otherwise be nine notifications a month for money
/// that leaves whether you are told or not.
bool reminderDue(Subscription sub, [DateTime? now]) {
  final want = sub.remindDays;
  if (want == null || want == 0) return false;
  final days = daysUntilRenewal(sub, now);
  if (days == null) return false;
  return days >= 0 && days <= want;
}

/// Everything that wants mentioning, soonest first.
List<Subscription> dueReminders(List<Subscription> subs, [DateTime? now]) {
  final out = subs.where((s) => reminderDue(s, now)).toList();
  out.sort((a, b) =>
      (daysUntilRenewal(a, now) ?? 0) - (daysUntilRenewal(b, now) ?? 0));
  return out;
}

/// Every renewal between two dates, inclusive.
///
/// Steps by asking `nextRenewal` again from the day after each hit rather than
/// doing its own arithmetic, so the end-of-month clamping is applied in
/// exactly one place. A weekly plan returns four or five dates in a month; a
/// yearly one returns nothing at all in eleven months of twelve, which is the
/// entire point of drawing this.
List<DateTime> renewalsBetween(Subscription sub, DateTime from, DateTime to) {
  final out = <DateTime>[];
  final last = startOfDay(to);
  var cursor = startOfDay(from);

  // 400 is a year and a half of weekly renewals — far past any window this is
  // called with, and a hard stop if a cadence ever fails to advance.
  for (var guard = 0; guard < 400; guard++) {
    final at = nextRenewal(sub, cursor);
    if (at == null || at.isAfter(last)) break;
    out.add(at);
    cursor = addDays(at, 1);
  }
  return out;
}

class MonthSpend {
  const MonthSpend(this.year, this.month, this.cents, this.count);
  final int year;

  /// 1-12. **The TypeScript stores 0-11**, because JavaScript months are
  /// zero-based; this is one of the deliberate differences in the port, and
  /// anything reading both has to know it.
  final int month;

  /// Real money, at real prices, on the days it actually leaves.
  final int cents;
  final int count;
}

/// What each of the next `months` calendar months actually costs.
///
/// A monthly total is an average, and an average hides the only thing about
/// subscription spending that ever surprises anybody: it is not level. Three
/// annual plans that happen to renew in January make January cost four times
/// what October does, and no amount of staring at "$94 a month" will tell you
/// that.
///
/// Whole months, including the part of this one already spent. A first bar
/// showing only what is left would be a different measurement from the five
/// beside it, which is the one thing a bar chart must never do.
List<MonthSpend> spendByMonth(List<Subscription> subs, int months,
    [DateTime? now]) {
  final at = now ?? DateTime.now();
  final out = <MonthSpend>[];

  for (var i = 0; i < months; i++) {
    final first = DateTime(at.year, at.month + i, 1);
    // Day zero of the next month is the last day of this one.
    final last = DateTime(first.year, first.month + 1, 0);

    var cents = 0;
    var count = 0;
    for (final s in subs) {
      final hits = renewalsBetween(s, first, last);
      count += hits.length;
      cents += hits.length * s.amountCents;
    }

    out.add(MonthSpend(first.year, first.month, cents, count));
  }

  return out;
}

/// The most expensive month in a run, or null if they are all empty.
MonthSpend? heaviest(List<MonthSpend> spend) {
  MonthSpend? top;
  for (final m in spend) {
    if (m.cents > 0 && (top == null || m.cents > top.cents)) top = m;
  }
  return top;
}
